tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/input_inline_fusion_pass.cc

```diff
@@ -0,0 +1,390 @@
+                                             .getElementType(),
+                                         operand_values, &rewriter);
+
+  for (auto* to_be_replaced : load_ops) {
+    llvm::cast<LoadOp>(to_be_replaced).replaceAllUsesWith(inlined_result);
+  }
+  return true;
+}
+
+template <typename LHLO_OpTy>
+bool miscFuseHelper(PatternRewriter& rewriter, Operation* user,
+                    Operation* producer, LoadOp load_op,
+                    std::vector<Operation*>& load_ops) {
+  if (!isa<LHLO_OpTy>(producer)) {
+    return false;
+  }
+  auto loc = user->getLoc();
+  rewriter.setInsertionPoint(load_op);
+  auto inlined_result =
+      elementalLower<LHLO_OpTy>(&rewriter, loc, producer, load_op.getIndices());
+  for (auto* to_be_replaced : load_ops) {
+    llvm::cast<LoadOp>(to_be_replaced).replaceAllUsesWith(inlined_result);
+  }
+  return true;
+}
+
+template <>
+bool miscFuseHelper<ConstOp>(PatternRewriter& rewriter, Operation* user,
+                             Operation* producer, LoadOp load_op,
+                             std::vector<Operation*>& load_ops) {
+  if (!isa<ConstOp>(producer)) {
+    return false;
+  }
+  auto memref_type =
+      cast<LmhloOp>(producer).getResultBuffer().getType().cast<MemRefType>();
+  auto rank = memref_type.getRank();
+  assert(rank == 0 && "only scalar ConstOp can be fused");
+  auto loc = user->getLoc();
+  rewriter.setInsertionPoint(load_op);
+  Value inlined_result =
+      rewriter.create<ConstantOp>(loc, memref_type.getElementType(),
+                                  cast<ConstOp>(producer).value().getValue({}));
+  for (auto* to_be_replaced : load_ops) {
+    llvm::cast<LoadOp>(to_be_replaced).replaceAllUsesWith(inlined_result);
+  }
+  return true;
+}
+
+template <typename First>
+bool elemwiseFuseHelperOr(PatternRewriter& rewriter, Operation* user,
+                          Operation* producer, LoadOp load_op,
+                          std::vector<Operation*>& load_ops) {
+  return elemwiseFuseHelper<First>(rewriter, user, producer, load_op, load_ops);
+}
+
+template <typename First, typename Second, typename... Rest>
+bool elemwiseFuseHelperOr(PatternRewriter& rewriter, Operation* user,
+                          Operation* producer, LoadOp load_op,
+                          std::vector<Operation*>& load_ops) {
+  return elemwiseFuseHelperOr<First>(rewriter, user, producer, load_op,
+                                     load_ops) ||
+         elemwiseFuseHelperOr<Second, Rest...>(rewriter, user, producer,
+                                               load_op, load_ops);
+}
+
+// load_op is among the load_ops, whose locates in the most
+// external code block
+LogicalResult InputInlineFusionPattern::inlineFuseLhloOp(
+    PatternRewriter& b, Operation* user, Operation* producer, LoadOp load_op,
+    std::vector<Operation*> load_ops) const {
+  if (elemwiseFuseHelperOr<
+#define GET_ELEMENTWISE_OP_LIST
+#include "tensorflow/compiler/mlir/hlo/lib/utils/elementwise_op_list.cc"
```

Comment:
Can you also upload `elementwise_op_list.cc`? I'm not sure which ops are defined as "elementwise" and interested to see. For example, whether reshapes and transposes in it?

Reply:
Uploaded.
Here elementwise means the same as XLA. The ops you don't need to calculate indices.
Reshape/Transpose is not 'elementwise'.